sdks/python/apache_beam/transforms/util.py

```diff
@@ -671,6 +679,60 @@ def WithKeys(pcoll, k):
   return pcoll | Map(lambda v: (k, v))
 
 
+@typehints.with_input_types(typehints.KV[K, V])
+class GroupIntoBatches(PTransform):
+  def __init__(self, batch_size):
+    self.batch_size = batch_size
+
+  @staticmethod
+  def of_size(batch_size):
+    return GroupIntoBatches(batch_size)
+
+  def expand(self, pcoll):
+    input_coder = coders.registry.get_coder(pcoll)
+    if not input_coder.is_kv_coder():
+      raise ValueError('coder specified in the input \
+      PCollection is not a KvCoder')
+    return pcoll | ParDo(_pardo_group_into_batches(
+        self.batch_size, input_coder))
+
+
+def _pardo_group_into_batches(batch_size, input_coder):
+  ELEMENT_STATE = BagStateSpec('values', input_coder)
+  COUNT_STATE = CombiningValueStateSpec('count', input_coder, CountCombineFn())
+  EXPIRY_TIMER = TimerSpec('expiry', TimeDomain.WATERMARK)
+
+  class _GroupIntoBatchesDoFn(DoFn):
+
+    def process(self, element,
```

Comment:
State/Timers does not work for all runners in python.

Instead, see this alternative implementation (https://github.com/apache/beam/blob/11aacd2ea2e2b1fec54efc0557e016e2b7f8d944/sdks/python/apache_beam/transforms/util.py#L430) Similar to the `_WindowAwareBatchingDoFn`, you batch elements in process and emit in process and finish_bundle.

Reply:
I'll take a look into this, but the ticket [https://issues.apache.org/jira/browse/BEAM-6696](https://issues.apache.org/jira/browse/BEAM-6696) says that it should use state to batch across bundles.

How should I interpret this ?

Thanks

Reply:
Good point.

We can use state/timers as you did here. Mark it as experimental and add a strongly worded comment or maybe even an actual warning to say that use of this requires state/timer support from the runner.

Reply:
fixed

Reply:
Thank you, resolving this.